File: scripts/oop-practice-sources/swift/user-auth.swift
class User {
    private let username: String
    private let password: String
    private var loggedIn = false

    init(username: String, password: String) {
        self.username = username
        self.password = password
    }

    func login(_ password: String) {
        if password == self.password {
            loggedIn = true
            print("Добро пожаловать, \(username)!")
        } else {
            print("Ошибка: неверный пароль")
        }
    }

    func logout() {
        loggedIn = false
        print("\(username) вышел из системы")
    }

    func postMessage(_ text: String) {
        if !loggedIn {
            print("Сначала войдите в систему")
            return
        }
        print("Сообщение опубликовано: \(text)")
    }
}

let user = User(username: "alex", password: "secret123")
user.postMessage("Привет!")
user.login("wrong")
user.login("secret123")
user.postMessage("Привет, мир!")
user.logout()
user.postMessage("Ещё одно сообщение")

File: scripts/oop-practice-sources/cpp/user-auth.cpp
#include <iostream>
#include <string>

class User {
private:
    std::string username;
    std::string password;
    bool logged_in;

public:
    User(const std::string& username, const std::string& password)
        : username(username), password(password), logged_in(false) {}

    void login(const std::string& pwd) {
        if (pwd == password) {
            logged_in = true;
            std::cout << "Добро пожаловать, " << username << "!" << std::endl;
        } else {
            std::cout << "Ошибка: неверный пароль" << std::endl;
        }
    }

    void logout() {
        logged_in = false;
        std::cout << username << " вышел из системы" << std::endl;
    }

    void post_message(const std::string& text) {
        if (!logged_in) {
            std::cout << "Сначала войдите в систему" << std::endl;
            return;
        }
        std::cout << "Сообщение опубликовано: " << text << std::endl;
    }
};

int main() {
    User user("alex", "secret123");
    user.post_message("Привет!");
    user.login("wrong");
    user.login("secret123");
    user.post_message("Привет, мир!");
    user.logout();
    user.post_message("Ещё одно сообщение");
    return 0;
}
